proposals/168/making_spark_visitors_extensible.md

```diff
@@ -0,0 +1,271 @@
+which determines the type of argument the builder can respond to. The `createInputDatasetBuilder` and 
+`createOutputDatasetBuilder` methods return `PartialFunction`s that construct either an `InputDatasetBuilder` or an
+`OutputDatasetBuilder`, respectively. Builders are returned so that facets constructed by the other `PartialFunction`s
+can be attached to the dataset before the immutable `InputDataset` or `OutputDataset` is returned. It is expected that
+the `CustomFacet`s returned are already immutable.
+
+For each `SparkListener` event, an `OpenLineage.RunEvent` will be constructed by 
+1. Construct `RunFacet`s by invoking each of the `runFacetBuilders`
+2. Construct `JobFacet`s by invoking each of the `jobFacetBuilders`
+3. Construct `InputDatasetBuilder`s by invoking each of the `inputDatasetBuilder`s.
+4. _If_ any `InputDatasetBuilder`s are returned, invoke the `inputDatasetFacetBuilder`s and attach any returned facets to
+   the `InputDatasetBuilder`s
+5. _If_ any `InputDatasetBuilder`s are returned, invoke the `datasetFacetBuilder`s and attach any returned facets to 
+    the `InputDatasetBuilder`s
+6. Construct `OutputDatasetBuilder`s by invoking each of the `outputDatasetBuilder`s.
+7. _If_ any `OuptutDatasetBuilder`s are returned, invoke the `outputDatasetFacetBuilder`s and attach any returned facets to
+   the `OutputDatasetBuilder`s
+5. _If_ any `OutputDatasetBuilder`s are returned, invoke the `datasetFacetBuilder`s and attach any returned facets to
+   the `OutputDatasetBuilder`s
+
+`InputDatasetFacet`s and `OutputDatasetFacet`s will be attached to _any_ `InputDatasetBuilder` or `OutputDatasetBuilder`
+found for the event. This is because facets may be constructed from generic information that is not specifi**cally tied
+to a Dataset. For example, `OutputStatisticsOutputDatasetFacet`s are created from `TaskMetrics` attached to** the last
+`StageInfo` for a given job execution. However, the `OutputDataset` is constructed by reading the `LogicalPlan`. There's
+no way to tie the output metrics in the `StageInfo` to the `OutputDataset` in the `LogicalPlan` except by inference.
+Similarly, input metrics can be found in the `StageInfo` for the stage that reads a dataset and the `InputDataset` can
+usually be constructed by walking the `RDD` dependency tree for that `Stage` and finding a `FileScanRDD` or other 
+concrete implementation. But while there is typically only one `InputDataset` read in a given stage, there's no 
+guarantee of that and the `TaskMetrics` in the `StageInfo` won't disambiguate. If a facet needs to be attached to a
+specific dataset, the user must take care to construct both the Dataset and the Facet in the same builder.
+
+For convenience, the following abstract base classes are defined:
+
+```java
+// accepts an argument of type T and returns an InputDatasetBuilder
+public abstract class AbstractInputDatasetBuilder<T>{}
```

Comment:
_Minor_: But is the `Abstract*` prefix redundant?

Reply:
This is pretty standard Java convention. E.g., `AbstractMap`, `AbstractList`, `AbstractCollection`, `AbstractFunction0`, `AbstractPartialFunction`, ... I think it helps readers of the concrete subclass to know that the subclass is extending an in complete base class without having to go look at the source code of the super class.